// egg-ts-helper/dist/generators/custom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const config_1 = require("../config");
const utils = tslib_1.__importStar(require("../utils"));
const path_1 = tslib_1.__importDefault(require("path"));
exports.isPrivate = true;
exports.defaultConfig = {
    directory: 'config',
    execAtInit: true,
    pattern: [
        'config(.local|.default|).(ts|js)',
        'plugin(.local|.default|).(ts|js)',
    ],
};
const customWatcherPrefix = 'custom-';
const DeclareMapping = utils.pickFields(config_1.declMapping, ['ctx', 'app']);
function default_1(config, baseConfig, tsHelper) {
    const createCustomLoader = (eggInfo) => {
        const eggConfig = eggInfo.config || {};
        const newCustomWatcherList = [];
        if (eggConfig.customLoader) {
            Object.keys(eggConfig.customLoader).forEach(key => {
                const loaderConfig = eggConfig.customLoader[key];
                if (!loaderConfig ||
                    !loaderConfig.directory ||
                    !DeclareMapping[loaderConfig.inject] ||
                    loaderConfig.tsd === false)
                    return;
                // custom d.ts name
                const name = `${customWatcherPrefix}${key}`;
                newCustomWatcherList.push(name);
                // create a custom watcher
                tsHelper.registerWatcher(name, {
                    distName: `${name}.d.ts`,
                    directory: loaderConfig.directory,
                    caseStyle: loaderConfig.caseStyle || 'lower',
                    interface: config_1.declMapping[key],
                    declareTo: `${DeclareMapping[loaderConfig.inject]}.${key}`,
                    generator: 'auto',
                    execAtInit: true,
                });
            });
        }
        // collect watcher which is need to remove.
        const removeList = [];
        tsHelper.watcherList.forEach(w => {
            if (w.name.startsWith(customWatcherPrefix) && !newCustomWatcherList.includes(w.name)) {
                removeList.push(w);
            }
        });
        // remove watcher and old d.ts
        return removeList.map(w => {
            tsHelper.destroyWatcher(w.name);
            return { dist: path_1.default.resolve(w.dtsDir, `${w.name}.d.ts`) };
        });
    };
    return utils.getEggInfo(baseConfig.cwd, {
        async: !!config.file,
        callback: createCustomLoader,
    });
}
exports.default = default_1;
